jumanji/environments/routing/rware/utils.py

```diff
@@ -0,0 +1,865 @@
+    sensor range.
+
+    Below is an example of the agent's view of other agents from
+    the perspective of agent 1 with a sensor range of 1:
+
+                            0, 0, 0
+                            0, 1, 2
+                            0, 0, 0
+
+    It sees agent 2 to its right. Separately, the view of shelves
+    is shown below:
+
+                            0, 0, 0
+                            0, 3, 4
+                            0, 7, 8
+
+    Agent 1 is on top of shelf 3 and has 4, 7 and 8 around it in
+    the bottom right corner of its view. Before returning these
+    views they are flattened into a 1-d arrays, i.e.
+
+    View of agents: [0, 0, 0, 0, 1, 2, 0, 0, 0]
+    View of shelves: [0, 0, 0, 0, 3, 4, 0, 7, 8]
+
+
+    Args:
+        grid: the warehouse floor grid array.
+        agent: the agent for which the view of their receptive field
+            is to be calculated.
+        sensor_range: the range of the agent's sensors.
+
+    Returns:
+        a view of the agents receptive field separated into two arrays:
+        one for other agents and one for shelves.
+    """
+    receptive_field = sensor_range * 2 + 1
+    padded_agents_layer = jnp.pad(grid[_AGENTS], sensor_range, mode="constant")
+    padded_shelves_layer = jnp.pad(grid[_SHELVES], sensor_range, mode="constant")
+    agent_view_of_agents = jax.lax.dynamic_slice(
+        padded_agents_layer,
+        (agent.position.x, agent.position.y),
+        (receptive_field, receptive_field),
+    ).reshape(-1)
+    agent_view_of_shelves = jax.lax.dynamic_slice(
+        padded_shelves_layer,
+        (agent.position.x, agent.position.y),
+        (receptive_field, receptive_field),
+    ).reshape(-1)
+    return agent_view_of_agents, agent_view_of_shelves
+
+
+def make_agent_observation(
+    grid: chex.Array,
+    agents: chex.Array,
+    shelves: chex.Array,
+    sensor_range: int,
+    num_obs_features: int,
+    highways: chex.Array,
+    agent_id: int,
+) -> chex.Array:
+    """Create an observation for a single agent based on its view
+    of other agents and shelves.
+
+    Args:
+        grid: the warehouse floor grid array.
+        agents: a pytree of either Agent type containing agent information.
+        shelves: a pytree of Shelf type containing shelf information.
+        sensor_range: the range of the agent's sensors.
+        num_obs_features: the number of features in the observation array.
+        highways: binary array indicating highway positions.
+        agent_id: unique ID identifying a specific agent.
+
+    Returns:
+        a 1-d array containing the agent's observation.
+    """
+    agent = tree_slice(agents, agent_id)
+    agents_grid, shelves_grid = get_agent_view(grid, agent, sensor_range)
+
+    # write flattened observations
+    obs = jnp.zeros(num_obs_features, dtype=jnp.int32)
+    idx = 0
+
+    # write current agent position and whether carrying a shelf or not
+    obs, idx = write_to_observation(
```

Comment:
I'm confused why this is necessary. If you know the size of your obs before hand could you not do: `obs = obs.at[idx:idx+len(data)].set(data)`? The `dynamic_update_slice` in `write_to_observation` isn't super common and I had to look that up, maybe this is more an issue with the `write_to_observation` method :thinking:

Reply:
Discussed and settled offline 👍